src/api/schema/filter.rs

```diff
@@ -0,0 +1,186 @@
+use super::components::{source, ComponentType};
+use async_graphql::{InputObject, InputType};
+
+/// Takes an &Option<bool> and returns early if false
+#[macro_export]
+macro_rules! filter_check {
+    ($($match:expr),+) => {
+        $(
+            if matches!($match, Some(t) if !t) {
+                return false;
+            }
+        )+
+    }
+}
+
+#[derive(Default, InputObject)]
+/// Filter for String values
+pub struct StringFilter {
+    pub equals: Option<String>,
+    pub not_equals: Option<String>,
+    pub contains: Option<String>,
+    pub not_contains: Option<String>,
+    pub starts_with: Option<String>,
+    pub ends_with: Option<String>,
+}
+
+impl StringFilter {
+    pub fn filter_value(&self, value: &str) -> bool {
+        filter_check!(
+            // Equals
+            self.equals.as_ref().map(|s| value.eq(s)),
+            // Not equals
+            self.not_equals.as_ref().map(|s| !value.eq(s)),
+            // Contains
+            self.contains.as_ref().map(|s| value.contains(s)),
+            // Does not contain
+            self.not_contains.as_ref().map(|s| !value.contains(s)),
+            // Starts with
+            self.starts_with.as_ref().map(|s| value.starts_with(s)),
+            // Ends with
+            self.ends_with.as_ref().map(|s| value.ends_with(s))
+        );
+        true
+    }
+}
+
+#[derive(InputObject)]
+#[graphql(concrete(name = "ComponentTypeFilter", params(ComponentType)))]
+#[graphql(concrete(name = "SourceOutputTypeFilter", params(source::SourceOutputType)))]
+pub struct EqualityFilter<T: InputType + PartialEq + Eq> {
+    pub equals: Option<T>,
+    pub not_equals: Option<T>,
+}
+
+impl<T: InputType + PartialEq + Eq> EqualityFilter<T> {
+    pub fn filter_value(&self, value: T) -> bool {
+        filter_check!(
+            // Equals
+            self.equals.as_ref().map(|s| value.eq(s)),
+            // Not equals
+            self.not_equals.as_ref().map(|s| !value.eq(s))
+        );
+        true
+    }
+}
+
+/// CustomFilter trait to determine whether to include/exclude fields based on matches.
+pub trait CustomFilter<T> {
+    fn matches(&self, item: &T) -> bool;
+    fn or(&self) -> Option<&Vec<Self>>
+    where
+        Self: Sized;
+}
+
+/// Returns true if a provided `Item` passes all 'AND' or 'OR' filter rules, recursively.
+fn filter_item<Item, Filter>(item: &Item, f: &Filter) -> bool
+where
+    Filter: CustomFilter<Item>,
+{
+    f.matches(item)
+        || f.or()
+            .map_or_else(|| false, |f| f.iter().any(|f| filter_item(item, f)))
```

Comment:
This feels like a good candidate for parallelism, a la [Rayon](https://github.com/rayon-rs/rayon) or similar.

I avoided it as messing with thread counts felt out of scope, but we might actually want to consider benchmarking `.par_iter` vs. `iter` in places. cc @jszwedko

Reply:
:+1: I'd suggest just setting up a micro benchmark around this function if you want to test `par_iter` vs. `iter`.